### PR TITLE
Fee provider dummy

### PR DESCRIPTION
**Pull request by cointigger**
# TLDR

A new **FeeProvider**: **DummyFeeProvider**

This fee provider is disabled by default and is configurable via configuration properties.

# Example configuration

* enable it explicitly
* two fees as map Duration(minutes) => Fee(sat/vByte)
  * 40sat/vByte for 30min
  * 30sat/vByte for 40min

```yaml
org:
  tbk:
    bitcoin:
      tool:
        fee:
          btcdotcom:
            enabled: false
          dummy:
            enabled: true
            static-fee-data: "30:40|40:30"
```

# Interesting changes

* **BitcoinFeeClientAutoConfigProperties**: ignoreUnknownFields changed to *true* - with *false* it is unable to contribute configuration entries ("Reason: The elements xxx were left unbound." -for ALL fee providers)
* **DummyFeeProviderAutoConfigProperties**: not linked to "org.tbk.bitcoin.tool.fee.enabled" as other providers (to enable this provider correctly)
* **versions.gradle**: added "org.apache.commons:commons-lang3:${apacheCommonsLang}" with apacheCommonsLong version in *versions.gradle*

### CODE REVIEW

File: spring-bitcoin-fee/spring-bitcoin-fee-autoconfigure/src/main/java/org/tbk/bitcoin/tool/fee/dummy/config/DummyFeeProviderAutoConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import org.springframework.context.annotation.Configuration;
+import org.tbk.bitcoin.tool.fee.dummy.ConfigurableDummyFeeSource;
+import org.tbk.bitcoin.tool.fee.dummy.DummyFeeProvider;
+import org.tbk.bitcoin.tool.fee.dummy.DummyFeeSource;
+
+import static java.util.Objects.requireNonNull;
+
+@Configuration
+@EnableConfigurationProperties(DummyFeeProviderAutoConfigProperties.class)
+@ConditionalOnProperty(name = "org.tbk.bitcoin.tool.fee.dummy.enabled", havingValue = "true")
```
Comment (cointigger) [inline spring-bitcoin-fee/spring-bitcoin-fee-autoconfigure/src/main/java/org/tbk/bitcoin/tool/fee/dummy/config/DummyFeeProviderAutoConfiguration.java:17]: Missing `@ConditionalOnClass` annotation here.

Without this a "ClassNotFound" may occurs if property set but module not implemented.

**Comment by theborakompanioni** [comment]
this is a really neat feature and generally something a user needs if a kind of fallback mechanism should step in. thank you so much for your contribution (this is the first pull request for this project - nice!).

some considerations when looking through the changes:
- it would be good if the dummy provider is ordered to be the last in line
- the configuration as string with delimiters is a little brittle.. maybe we can deserialize it as `Map<Duration, BigDecimal>`?
  (i do not know if this is really applicable).. e.g. 
  ```yaml
    - 'PT0S': '10'
    - 'PT30M': '5'
    - 'PT6H': '3'
    - 'PT24H': '1'
  ```
- i guess maybe it is desired to be able to change these values during runtime (otherwise the application would need to be restarted). It should be possible to expose the object as jmx bean.

What are your thoughts on this?

**Comment by cointigger** [comment]
> this is a really neat feature and generally something a user needs if a kind of fallback mechanism should step in. thank you so much for your contribution (this is the first pull request for this project - nice!).
> 
> some considerations when looking through the changes:
> 
>     * it would be good if the dummy provider is ordered to be the last in line
> 
>     * the configuration as string with delimiters is a little brittle.. maybe we can deserialize it as `Map<Duration, BigDecimal>`?
>       (i do not know if this is really applicable).. e.g.
>       ```yaml
>         - 'PT0S': '10'
>         - 'PT30M': '5'
>         - 'PT6H': '3'
>         - 'PT24H': '1'
>       ```
> 
>     * i guess maybe it is desired to be able to change these values during runtime (otherwise the application would need to be restarted). It should be possible to expose the object as jmx bean.
> 
>     * is apache commons only pulled in to use the `Pair` class? I'd generally want to avoid too much dependencies, especially when it is only used in private methods - maybe it can be avoided.
> 
> 
> What are your thoughts on this?

**Comment by cointigger** [comment]
>     * it would be good if the dummy provider is ordered to be the last in line

Good idea, I kept the order as I found it: alphabetically. But move this provider to the end is a good idea.

>     * the configuration as string with delimiters is a little brittle.. maybe we can deserialize it as `Map<Duration, BigDecimal>`?
>       (i do not know if this is really applicable).. e.g.
>       ```yaml
>         - 'PT0S': '10'
>         - 'PT30M': '5'
>         - 'PT6H': '3'
>         - 'PT24H': '1'
>       ```

I'm not so familiar with all fancy capabilities of spring config - yes, a list as mentioned is way better to manage as my simple String approach. Good point.

>     * i guess maybe it is desired to be able to change these values during runtime (otherwise the application would need to be restarted). It should be possible to expose the object as jmx bean.

>     * is apache commons only pulled in to use the `Pair` class? I'd generally want to avoid too much dependencies, especially when it is only used in private methods - maybe it can be avoided.

The algorithm "split string into records, and split each record into a pair of data" was the reason to include it, yes.

For sure it can be replaces easily, I'm afraid I was overwhelmed by the power of module-own dependencies.

Maybe with a smarter config as mentioned above the need of `Pair` is not longer active.

Will change it.

Thx for your suggestions!

**Comment by cointigger** [comment]
Suggestions added except this one:

> i guess maybe it is desired to be able to change these values during runtime (otherwise the application would need to be restarted). It should be possible to expose the object as jmx bean.

Unfortunately, I'm not aware with the magic of JMX at all - so expose a runtime refresh mechanism is not part of this solution.
